dynamiqs/plots/utils.py

```diff
@@ -0,0 +1,147 @@
+from __future__ import annotations
+
+from functools import wraps
+
+import matplotlib as mpl
+import numpy as np
+from matplotlib import pyplot as plt
+from matplotlib.axes import Axes
+from matplotlib.axis import Axis
+from matplotlib.cm import get_cmap
+from matplotlib.colors import Normalize
+from matplotlib.figure import Figure
+from matplotlib.ticker import MultipleLocator, NullLocator
+from mpl_toolkits.axes_grid1 import make_axes_locatable
+
+__all__ = [
+    'figax',
+    'optax',
+    'mplstyle',
+    'integer_ticks',
+    'sample_cmap',
+    'minorticks_off',
+    'fock_ticks',
+    'add_colorbar',
+]
+
+
+def figax(w: float = 10.0, h: float | None = None, **kwargs) -> tuple[Figure, Axes]:
+    """Return a figure with specified width and length."""
+    if h is None:
+        h = w / 2
+    return plt.subplots(1, 1, figsize=(w, h), **kwargs)
+
+
+def optax(func):
+    """Decorator to build an `Axes` object to pass as an argument to a plot
+    function if it wasn't passed by the user.
+
+    Examples:
+        Replace
+        ```
+        def myplot(ax=None):
+            if ax is None:
+                _, ax = plt.subplots(1, 1)
+
+            # ...
+        ```
+        by
+        ```
+        @optax
+        def myplot(ax=None):
+            # ...
+        ```
+    """
+
+    @wraps(func)
+    def wrapper(
+        *args,
+        ax: Axes | None = None,
+        w: float = 10.0,
+        h: float | None = None,
+        **kwargs,
+    ):
+        if ax is None:
+            _, ax = figax(w=w, h=h)
+        return func(*args, ax=ax, **kwargs)
+
+    return wrapper
+
+
+def mplstyle(*, latex: bool = True):
+    """Set custom Matplotlib style."""
+    plt.rcParams.update(
+        {
+            'figure.facecolor': 'white',
+            'axes.facecolor': 'white',
+            'font.size': 12,
+            'figure.dpi': 72.0,
+            'xtick.direction': 'in',
+            'ytick.direction': 'in',
+            'xtick.major.size': 5.0,
+            'xtick.minor.size': 2.5,
+            'ytick.major.size': 5.0,
+            'ytick.minor.size': 2.5,
+            'xtick.minor.visible': True,
+            'ytick.minor.visible': True,
+            'axes.grid': False,
+            'axes.titlesize': 'larger',
+            'axes.labelsize': 'larger',
+            'grid.color': 'gray',
+            'grid.linestyle': '--',
+            'grid.alpha': 0.3,
+            'scatter.marker': 'x',
+            'lines.linewidth': 1.0,
+        }
+    )
```

Comment:
```suggestion
    plt.rcParams.update(
        {
        'figure.facecolor': 'white',
        'axes.facecolor': 'white',
        'savefig.facecolor': 'white',
        'font.size': 12,
        'figure.dpi': 150,
        'xtick.direction': 'in',
        'ytick.direction': 'in',
        'xtick.major.size': 4.0,
        'xtick.minor.size': 2.5,
        'xtick.major.width': 1.0,
        'xtick.labelsize': 10,
        'xtick.minor.visible': True,
        'ytick.major.size': 4.0,
        'ytick.minor.size': 2.5,
        'ytick.major.width': 1.0,
        'ytick.labelsize': 10,
        'ytick.minor.visible': True,
        'axes.grid': False,
        'axes.titlesize': 12,
        'axes.labelsize': 12,
        'axes.linewidth': 1.0,
        'axes.prop_cycle': cycler('color', ['#1398c4', '#ff6b6b', '#2ec4b6', '#ffc463', '#869bac', '#966f94', '#c0675c', '#20817d']),
        'grid.color': 'gray',
        'grid.linestyle': '--',
        'grid.alpha': 0.3,
        'scatter.marker': 'x',
        'lines.linewidth': 1.5,
        'legend.frameon': False,
        'legend.fontsize': 10,
        'legend.borderpad': 0.5,
        'legend.labelspacing': 0.08,
        'legend.handlelength': 1.2,
        'legend.handletextpad': 0.4,
        'legend.columnspacing': 0.5,
        'figure.figsize': [4.0, 2.5],
        }
    )
```
(taken from my own mplstyle)

Reply:
```python
import matplotlib.pyplot as plt
from cycler import cycler
import numpy as np

plt.rcParams.update({})

# sample data
n = 16
x = np.linspace(0, 10, 30)
y = np.zeros((30, n))
for i in range(n):
    y[:, i] = 0.3 * i + np.sin(x)

# sample plot
fig, ax = plt.subplots(1, 1)
ax.plot(x, y, lw=2)
ax.set_xlabel('Time')
ax.set_ylabel('Sine')
ax.set_title('Sample plot')
plt.tight_layout()
plt.show()
```
![image](https://github.com/dynamiqs/dynamiqs/assets/25346881/2747e96e-74cf-4680-a2ef-272919114482)
![image](https://github.com/dynamiqs/dynamiqs/assets/25346881/482e67d1-f6b3-42e2-8235-06931c9e4c3d)

Reply:
Feel free to modify it whichever way you like in case you want to adopt this styling. If you don't, that's fine with me as well.

Reply:
Thanks a lot for sharing!

I took everything except:
```python
'figure.dpi': 150   # -> kept 72 for now
'xtick.major.size': 4.0  # -> 4.5
'xtick.labelsize': 10  # -> 12
'ytick.major.size': 4.0  # -> 4.5
'ytick.labelsize': 10  # -> 12
'legend.fontsize': 10  # -> 12
'legend.labelspacing': 0.08  # -> removed
'legend.handlelength': 1.2  # -> removed
'legend.handletextpad': 0.4  # -> removed
'legend.columnspacing': 0.5  # -> removed
'lines.linewidth': 1.5,  # -> 2.0
'figure.figsize': [4.0, 2.5]  # -> removed
```

Here are the two styles:
![2fb88817-9eed-4a4c-8323-3a860b29aaf2](https://github.com/dynamiqs/dynamiqs/assets/38159029/5aa216a3-565f-4374-9923-73cd7529d0d5)

![dc073e9a-36be-4507-aa90-cf1b67a0f3fc](https://github.com/dynamiqs/dynamiqs/assets/38159029/aebc516e-a663-494b-92ea-f55924636eae)

I think the first plot looks super nice, but is better suited for a research paper. I like having large detailed plots when working. I think we should propose both styles actually, and let people choose. For the colors I did a little mix-and-match so my current plot look nice, but I think we need to iterate too 😅 It might be better to write 2-3 `.mplstyle` file. I propose to merge this as is, and iterate on the styling later.

Reply:
I think the best is to iterate on the styling once we have a representative set of figures that we plot often (fock, wigner, pulse, etc...) to make sure the choices we make are well suited to all basic figures.

Reply:
Agreed. All good !